Flow HttpClient timeouts and cancellation to the response stream

stephentoub opened the pull request:
When using a timeout and/or a cancellation token with a buffered HttpClient operation, the cancellation/timeout doesn't flow through to the actual read/write operations on the underlying streams.  This fixes that.

cc: @davidsh, @cipop, @ericeil, @bartonjs 
Fixes #8663

davidsh commented on src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs:
```diff
@@ -15,110 +15,108 @@ namespace System.Net.Http.Functional.Tests
+                            $"Date: {DateTimeOffset.UtcNow:R}\r\n" +
+                            "Content-Length: 16000\r\n" +
+                            "\r\n" +
+                            "less than 16000 bytes");
```
Switching to a loopback server was a TODO that I had.  But this change is more than just using a loopback server. The intent is the tests was to have response headers sent right away but then have response body payload arrive later ("FastHeadersSlowBody").

This change using the LoopbackServer is fundamentally different.  It sends the response headers and 99% of the payload out right away. It's "99%" because the test code is intentionally sending out less bytes than the 'Content-Length'.  While this is an interesting test, it does not exercise the same codepaths in WinHttpHandler, for example, that I was explicitly testing against. I.e. I was testing for the WinHttpQueryDataAvailable() call taking a long time because there were 0 bytes available.  So, this change here makes this a totally different kind of test.

stephentoub replied:
> The intent is the tests was to have response headers sent right away but then have response body payload arrive later ("FastHeadersSlowBody").

Ah. Ok.  It's easy enough to change.  We can just reorganize this part slightly, for example removing this few bytes of body here and putting them in their own write after the ```await triggerResponseWrite.Task```.

I assume both cases are interesting, though, so we can we can just augment the theory to test both.

davidsh replied:
>I assume both cases are interesting, though, so we can we can just augment the theory to test both.

Using a theory is good.

bartonjs commented:
Main code LGTM; I didn't really look at the tests since there was already a discussion going on there.

CIPop commented on src/System.Net.Http/src/System/Net/Http/StreamContent.cs:
```diff
@@ -12,19 +11,30 @@ namespace System.Net.Http
+        {
+        }
+
+        internal StreamContent(Stream content, CancellationToken cancellationToken)
```
/cc @davidsh, @himadrisarkar 
Since the other ctor is public, would it be useful to add this as a public API?

davidsh replied:
>new StreamContent(decompressedStream, state.CancellationToken);

It's rather un-intuitive to pass in a cancellation token to a constructor like StreamContent. What operation needs to be cancelled? So, I do not think this ctor overload should be public at all. I have some other ideas I'm thinking about, to flow the token to the real operation which is the StreamToStream copy operation.

stephentoub replied:
> What operation needs to be cancelled?

The token needs to get into the StreamToStream copy operation, but the virtual SerializeToStreamAsync method doesn't take a CancellationToken.

davidsh replied:
There is still time to adjust the 4.1 System.Net.Http contract.  I probably should have done it earlier.  But the SerializeToStreamAsync and CopyToAsync methods should have overloads to take a CancellationToken.  The SerializeToStreamAsync method is abstract and adding another abstract method to the HttpContent base class isn't allowed. But we could add a new virtual method overload of SerializeToStreamAsync to take a token. And the other *Async methods could take one as well.  That would be the best way to pass in a token.

stephentoub replied:
> That would be the best way to pass in a token.

Yup, agreed.  This ctor approach is a temporary workaround for the lack of such APIs.

I suggest we go ahead with this change, and then this small piece of it can be undone when such overloads are available; if nothing else, it'll serve as a good test case to help validate the design.

davidsh replied:
>I suggest we go ahead with this change

Ok. We can defer the new overloads to a 4.2 System.Net.Http API (which I am already planning for) in the next few months.

stephentoub replied:
> We can defer the new overloads to a 4.2 System.Net.Http API (which I am already planning for) in the next few months.

To be clear, I wasn't suggesting we not try to do something for the 4.1 contract, just that we not hold up this PR on that.  It'd be good to at least sketch out what the new overloads would be and how we see them interplaying with existing code.  Depending on how its done, it could be problematic.  For example, let's say you have a method:
```
protected virtual Task Foo();
```
which the consumer calls.  Now we add a new overload:
```
protected virtual Task Foo(CancellationToken cancellationToken);
```
For the newly added method to be useful, we would need to switch to calling it instead of calling the former.  But no existing implementations will be overriding it, which means for compatibility the base implementation would need to delegate to the old implementation, at which point the CancellationToken is ignored.

davidsh replied:
In terms of response body buffering, the *Content object is always an `StreamContent` object.  So, we have total control of that.

Developers create custom HttpContent-derived classes for the purpose of handling request body content. I.e., to make it easier for them to bundle up that content to POST, instead of having to serialize it themselves as ByteArrayContent, for example.

But this scenario of being able to cancel response body buffering only uses the StreamContent class which is part of the System.Net.Http library.

stephentoub replied:
Ok, so when the new overloads are added, StreamContent's overrides of the new overloads will continue to delegate to the old overloads, but we can introduce a new internal ResponseStreamContent type that derives from StreamContent and overrides the new overload to do the right thing.

davidsh replied:
>Ok, so when the new overloads are added, StreamContent's overrides of the new overloads will continue to delegate to the old overloads, but we can introduce a new internal ResponseStreamContent type that derives from StreamContent and overrides the new overload to do the right thing.

Yes.

CIPop commented:
There is a compilation problem:

```
System\Net\Http\WinHttpResponseParser.cs(86,65): error CS1503: Argument 2: cannot convert from 'System.Threading.CancellationToken' to 'int' [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Http.WinHttpHandler\src\System.Net.Http.WinHttpHandler.csproj]
14:07:41 
System\Net\Http\WinHttpResponseParser.cs(86,65): error CS1503: Argument 2: cannot convert from 'System.Threading.CancellationToken' to 'int' [D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Http.WinHttpHandler\src\System.Net.Http.WinHttpHandler.csproj]
```

The WinHttpHandler unit and functional tests may need to change as well?

stephentoub commented:
Test Innerloop CentOS7.1 Debug Build and Test please (https://github.com/dotnet/corefx/issues/9053)
Test Innerloop CentOS7.1 Release Build and Test please (https://github.com/dotnet/corefx/issues/9053)

stephentoub commented:
I updated the PR to address the compilation error and to incorporate David's feedback on the test to test more cases (no response body sent vs some response body sent).

For the compilation error, the issue is that System.Net.Http.WinHttpHandler.dll pulls in the same source file, but it doesn't have access to the internal ctor I added on StreamContent.  Without allowing StreamContent to know in some way what CancellationToken to use, there isn't a great way to enable this.  For now, based on discussing with @cipop, I've ifdef'd it, so HttpClientHandler will get the better behavior and WinHttpHandler will remain as-is.  I'll open a tracking issue and include the issue number at https://github.com/dotnet/corefx/pull/9041/files#diff-933c557c2433cae9802bf40cf9b4ffbfR89 before merging, unless someone has a better suggestion for how to handle this.

davidsh commented:
LGTM

stephentoub commented:
Test Outerloop Windows 10 Debug please
Test Outerloop Ubuntu14.04 Debug please

stephentoub commented:
Test Innerloop Ubuntu14.04 Debug Build and Test please (out of disk space)

CIPop commented:
LGTM Thanks @stephentoub !

stephentoub commented:
Test OuterLoop Windows 10 Debug please

stephentoub commented:
Test Innerloop Ubuntu14.04 Debug Build and Test please

stephentoub commented:
Test OuterLoop Windows 10 Debug please

stephentoub commented:
Test OuterLoop Ubuntu14.04 Debug please

stephentoub commented:
Test OuterLoop Windows 10 Debug please (Jenkins isn't responding to these requests; trying again)

stephentoub commented:
(Closing and reopening this PR; Jenkins isn't recognizing it.)
